### PR TITLE
feat: integrate Yarn workspaces with Lerna for smoother DX without false positive type errors

### PR DESCRIPTION
**Pull request by grabbou**
Removes deprecated `lerna bootstrap` (and its own package manager) in favour of Yarn workspaces. As a benefit, we're finally getting an error-free developer experience where VSCode (and other editors) can properly resolve monorepo packages (such as `@web3-react/types` and provide good DX).

To better visualise, this PR fixes the following errors:

Screenshot 2023-02-04 at 02 29 56
Screenshot 2023-02-04 at 02 29 50

Thanks to that, we can develop faster without overlooking potentially dangerous type errors.

This PR also:
- removes `example-cra` package that doesn't even install successfully with Yarn workspaces (and we're not maintaining it either)
- fixes few small type issues that I found after fixing aforementioned type errors
- removes unnecessary and obsolete readme instructions for managing dependencies (it's a simple yarn install now)
- moves the example folder to the root of the repository

Note that due to the nature of Yarn workspaces, multiple project-specific lock files are now removed in favour of a single top-level lock-file. You will notice that it's mostly Next.js, since other packages were already peer dependencies on a top-level workspace package.

### CODE REVIEW

**Comment by grabbou** [comment]
Note that Vercel failures are expected as the configuration is hosted in the cloud and applies to all branches. Once this PR is merged, I will update the cloud configuration so that they continue to work again.

PS. Long term wise, this should be defined in a branch, instead of a cloud dashboard.

File: packages/walletconnect/tsconfig.json [hunk 1]
```diff
@@ -2,6 +2,7 @@
   "extends": "../../tsconfig.json",
   "include": ["./src"],
   "compilerOptions": {
-    "outDir": "./dist"
+    "outDir": "./dist",
+    "esModuleInterop": true
```
Comment (grabbou) [inline packages/walletconnect/tsconfig.json:6]: Typescript complained this was required for the `eventemitter3`. This is interesting, because pre and post my work, the version is the same. `tsc` is actually not doing type checking while compiling, and I guess that's the reason why it worked. Now that all the remaining type errors are gone, I have updated this configuration to make it have no errors at all.

**Review by cmcewen — COMMENTED** [review]
my main concern is just about any unexpected dependency changes - is there any reason that removing all the yarn.lock files from the individual packages would lead to different resolutions? also maybe just verifying that publishing would work?

File: example/components/ConnectWithSelect.tsx [hunk 2]
```diff
@@ -136,7 +136,7 @@ export function ConnectWithSelect({
         <div style={{ marginBottom: '1rem' }} />
         <button
           onClick={() => {
-            if (connector?.deactivate) {
+            if ('deactivate' in connector) {
```
[the remark this answers is not in the document]
Comment (grabbou) [reply example/components/ConnectWithSelect.tsx]: used to trigger an error while I was working on it. Surprisingly, the error is now gone, so we're good. Thanks for pointing it out.

File: package.json [hunk 3]
```diff
@@ -5,17 +5,20 @@
...
   "scripts": {
-    "bootstrap": "lerna bootstrap",
     "clean": "lerna clean --yes",
     "lint": "yarn run eslint --ext .ts,.tsx .",
     "test": "jest",
     "build": "lerna run build",
     "prestart": "yarn build",
     "start": "lerna run start --parallel",
-    "pin": "lerna run pin && yarn bootstrap",
+    "pin": "lerna run pin && yarn",
```
[the remark this answers is not in the document]
Comment (grabbou) [reply package.json:14]: Makes sense, less confusing when it stays on the package level.

File: packages/types/src/index.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import type { EventEmitter } from 'node:events'
-import type { State, StoreApi } from 'zustand'
+import type { StoreApi } from 'zustand'
 
-export interface Web3ReactState extends State {
```
Comment (cmcewen) [inline packages/types/src/index.ts:4]: also why this change?

Comment (grabbou) [reply packages/types/src/index.ts:4]: > An interface can only extend an object type or intersection of object types with statically known members.ts(2312)

+ 

> 'State' is deprecated. Use `unknown` instead of State.

which in our case means that we can simply drop it.

**Comment by grabbou** [comment]
> my main concern is just about any unexpected dependency changes - is there any reason that removing all the yarn.lock files from the individual packages would lead to different resolutions? also maybe just verifying that publishing would work?

Theoretically, it could change. However, most of the sub-package yarn.locks were rather minimal, since we don't have much dependencies. Most of the changes are caused by an example, which is an example and I don't think it matters there that much. However, I copied its yarn.lock manually to the root yarn.lock before running yarn to make sure that the versions are picked up correctly. I then run `npx yarn-deduplicate`.

File: packages/types/src/index.ts (hunk 4 above)
Comment (cmcewen) [reply packages/types/src/index.ts:4]: i'm guessing this was only surfaced because of fixing the type errors. i'm okay with this one going in here, but it does help to review code changes vs infra changes separately as much as possible

**Comment by cmcewen** [comment]
looks like we need to change the vercel deploy command - fails on bootstrap

File: packages/types/src/index.ts (hunk 4 above)
Comment (grabbou) [reply packages/types/src/index.ts:4]: I did run into this error in #750 too, if you look into the diff, it is fixed there as well. My goal for this PR was to ensure there are no extraneous type errors, hence I decided to apply those changes in a single batch. We can split it for the next time!

**Comment by grabbou** [comment]
@cmcewen yes, I talked to @zzmp and we agreed to change it post merge, so that main is not affected https://github.com/Uniswap/web3-react/pull/752#issuecomment-1423009034

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (grabbou) [reply package.json]: I ended up removing `pin` entirely and using `lerna` instead. Why `lerna` works this time and didn't work before?

I added a version to `example-next`, which made it a valid package to be picked up by Lerna when doing a versioning. I tried it locally and can confirm that it works like a charm - all versions are in check after running `yarn lerna version`.

**Review by cmcewen — APPROVED** [review]
🚢
